### PR TITLE
Fixes #196: Create option for whitelisting folders in the settings

### LINKED ISSUE #196: Create option for whitelisting folders in the settings
**Actual Behaviour**

Currently the app scans through the phone's folders to get all the images.

**Expected Behaviour**

The users will be able to whitelist folders in the settings to avoid those folders being scanned.

**Would you like to work on the issue?**

Yes, I would like to work on it.

### PR DESCRIPTION
**Pull request by rohanagarwal94**
Fixes issue #196

Currently, the app iterates through the images of the app with the help of the MediaStore class to get the images. Hence the user does not have the option to choose the folder in that case. Therefore, the cursor implementation needs to be changed wherever the image grid is being shown. I will be creating separate issue for that since that is altogether a time consuming task itself. The rest of the implementation is included in the PR.

Screenshots for the change: 

screenshot_2017-04-02-04-28-06-286_vn mbm phimp me
screenshot_2017-04-02-04-10-03-370_vn mbm phimp me
screenshot_2017-04-02-04-09-54-671_vn mbm phimp me

### CODE REVIEW

**Comment by phimpme-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">:tada: Thanks for the PR @rohanagarwal94 we hope you have changed the variables in the upload_apk_forks.sh its for easy moderation.</td>
    </tr>
  </tbody>
</table>

:link: Maintainers: the apk to test can be found at https://github.com/rohanagarwal94/phimpme-android/blob/apk/app-debug.apk

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by rohanagarwal94** [comment]
@pa1pal @Abhi2424shek any reason that the travis build failed?

**Comment by Abhi2424shek** [comment]
Espresso tests are failing that's why the build is failing. @rohanagarwal. Also there are conflicts please try to resolve them as well.

**Comment by rohanagarwal94** [comment]
@pa1pal @Abhi2424shek please review. Resolved all conflicts.

File: app/src/main/java/vn/mbm/phimp/me/PhimpMe.java [hunk 1]
```diff
@@ -857,56 +862,60 @@ private void changeRtoL() {
...
+//                        PhimpMe.cache.saveCacheFile(path, bmp, id);
+//                    } else id = -1;
+//
+//                    c++;
+//                }
+//
+//            }
+//            newGallery.update_number++;
+//
+//        }
```
Comment (Abhi2424shek) [inline app/src/main/java/vn/mbm/phimp/me/PhimpMe.java]: Please remove unused code.

File: app/src/main/java/vn/mbm/phimp/me/folderchooser/AppState.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * Created by rohanagarwal94 on 1/4/17.
+ */
+
+public class AppState {
+    private List<FileTree> scannedFileTree;
+    private List<StorageInfo> storageInfos;
+
+    private static AppState ourInstance = new AppState();
+
+    private AppState() {
```
Comment (Abhi2424shek) [inline app/src/main/java/vn/mbm/phimp/me/folderchooser/AppState.java]: Unused constructor?

File: app/src/main/java/vn/mbm/phimp/me/folderchooser/FileTree.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        this.parent = parent;
+    }
+
+    public void hasMedia() {
+        hasMedia = true;
+        if (parent != null) {
+            parent.hasMedia();
+        }
+    }
+}
```
Comment (Abhi2424shek) [inline app/src/main/java/vn/mbm/phimp/me/folderchooser/FileTree.java]: Add a newline at eof

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -641,9 +641,16 @@
...
+    <string name="folder_settings">Folder Settings</string>
+    <string name="whitelist_folders">WhiteList Folders</string>
+    <string name="whitelist_folders_message">Click to WhiteList Folders</string>
+
     <!-- edit Image Dialog -->
     <string name="go_back">Go Back</string>
     <string name="dialog_message">Do you want to exit and discard changes?</string>
     <string name="dialog_title">Confirm</string>
-
+
```
Comment (Abhi2424shek) [inline app/src/main/res/values/strings.xml]: Trailing whitespace

**Comment by rohanagarwal94** [comment]
@Abhi2424shek @pa1pal made the changes, please review.

**Comment by mariobehling** [comment]
Thanks a lot! Keep it up. Looking forward to more contributions from you.

**Pull request merged**